src/include/daos_pipeline.h

```diff
@@ -0,0 +1,261 @@
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+
+/** logic functions */
+#define DAOS_FILTER_FUNC_EQ		0
+#define DAOS_FILTER_FUNC_NE		1
+#define DAOS_FILTER_FUNC_LT		2
+#define DAOS_FILTER_FUNC_LE		3
+#define DAOS_FILTER_FUNC_GE		4
+#define DAOS_FILTER_FUNC_GT		5
+#define DAOS_FILTER_FUNC_LIKE		6
+#define DAOS_FILTER_FUNC_ISNULL		7
+#define DAOS_FILTER_FUNC_ISNOTNULL	8
+#define DAOS_FILTER_FUNC_AND		9
+#define DAOS_FILTER_FUNC_OR		10
+
+/** aggregation functions */
+#define DAOS_FILTER_FUNC_SUM		100
+#define DAOS_FILTER_FUNC_MIN		101
+#define DAOS_FILTER_FUNC_MAX		102
+#define DAOS_FILTER_FUNC_AVG		103
+
+/** keys, constants */
+#define DAOS_FILTER_DKEY		200
+#define DAOS_FILTER_AKEY		201
+#define DAOS_FILTER_CONST		202
+
+/** types used in a filter object */
+#define DAOS_FILTER_TYPE_BINARY		0
+#define DAOS_FILTER_TYPE_STRING		1
+#define DAOS_FILTER_TYPE_INTEGER	2
+#define DAOS_FILTER_TYPE_REAL		3
+
+/** types of pipeline nodes */
+#define DAOS_PIPELINE_CONDITION		0
+#define DAOS_PIPELINE_AGGREGATION	1
+
+
+/**
+ * A filter object, used to build operations for a pipeline node.
+ *
+ */
+typedef struct {
+	/**
+	 * filter type can be any of the following:
+	 *   -- functions:
+	 *      - logical functions:
+	 *          DAOS_FILTER_FUNC_EQ:		==
+	 *          DAOS_FILTER_FUNC_NE:		!=
+	 *          DAOS_FILTER_FUNC_LT:		<
+	 *          DAOS_FILTER_FUNC_LE:		<=
+	 *          DAOS_FILTER_FUNC_GE:		>=
+	 *          DAOS_FILTER_FUNC_GT:		>
+	 *          DAOS_FILTER_FUNC_LIKE:		== (reg exp.)
+	 *          DAOS_FILTER_FUNC_ISNULL:		==NULL
+	 *          DAOS_FILTER_FUNC_ISNOTNULL:		!=NULL
+	 *          DAOS_FILTER_FUNC_AND:		&&
+	 *          DAOS_FILTER_FUNC_OR:		||
+	 *      - aggeration functions:
+	 *          DAOS_FILTER_FUNC_SUM:		SUM()
+	 *          DAOS_FILTER_FUNC_MIN:		MIN()
+	 *          DAOS_FILTER_FUNC_MAX:		MAX()
+	 *          DAOS_FILTER_FUNC_AVG:		AVG()
+	 *   -- key:
+	 *          DAOS_FILTER_DKEY:	Filter object represents dkey
+	 *          DAOS_FILTER_AKEY	Filter object represents akey
+	 *   -- constant:
+	 *          DAOS_FILTER_CONST:	Filter object is a constant
+	 */
+	int		filter_type;
+	/**
+	 * Type of data. Only relevant for keys and constant filter type objects:
+	 *          DAOS_FILTER_TYPE_BINARY
+	 *          DAOS_FILTER_TYPE_STRING
+	 *          DAOS_FILTER_TYPE_INTEGER
+	 *          DAOS_FILTER_TYPE_REAL
+	 */
+	int		data_type;
+	/**
+	 * Number of parameters for this filter object. For example, for '=='
+	 * we have 2 parameters.
+	 */
+	uint32_t	num_params;
+	/**
+	 * If filtering by akey, this tells us which one.
+	 */
+	d_iov_t		akey;
+	/**
+	 * How many constants we have in \a constant
+	 */
+	size_t		num_constants;
+	/**
+	 * This object holds the value of the constants
+	 */
+	d_iov_t		*constant;
+	/**
+	 * If filter should only be applied starting at an offset of the data.
+	 */
+	size_t		data_offset;
+	/**
+	 * Size of the data to be filtered.
+	 */
+	size_t		data_len;
+} daos_pipeline_filter_t;
```

Comment:
This is actually just one filter, right? So why not just daos_filter_t?

Reply:
The reason is to emphasize the context. That is, it is a filter for a pipeline. But we can surely change it if it is too long.

Reply:
This is actually just the part of a filter. It will be renamed daos_filter_part_t